source/crud/bulk-write.md

```diff
@@ -0,0 +1,820 @@
+    "upsert": Optional<Boolean>,
+    "arrayFilters": Optional<Array>,
+    "hint": Optional<Document | String>
+}
+```
+
+#### Delete
+
+```json
+{
+    "delete": <Int32>,
+    "filter": <Document>,
+    "multi": Optional<Boolean>,
+    "hint": Optional<Document | String>,
+    "collation": Optional<Document>
+}
+```
+
+### Namespace Information
+
+The `nsInfo` field is an array containing the namespaces on which the write operations should be
+performed. Drivers MUST NOT include duplicate namespaces in this list. The documents in the
+`nsInfo` array have the following format:
+
+```json
+{
+    "ns": <String>
+}
+```
+
+### `errorsOnly` and `verboseResults`
+
+The `errorsOnly` field indicates whether the results cursor returned in the `bulkWrite` response
+should contain only errors and omit individual results. If false, both individual results for
+successful operations and errors will be returned. This field is optional and defaults to false on
+the server.
+
+`errorsOnly` corresponds to the `verboseResults` option defined on `BulkWriteOptions`. If the user
+specified a value for `verboseResults`, drivers MUST define `errorsOnly` as the opposite of
+`verboseResults`. If the user did not specify a value for `verboseResults`, drivers MUST define
+`errorsOnly` as `true`.
+
+### `ordered`
+
+The `ordered` field defines whether writes should be executed in the order in which they were
+specified, and, if an error occurs, whether the server should halt execution of further writes. It
+is optional and defaults to true on the server. Drivers MUST explicitly define `ordered` as `true`
+in the `bulkWrite` command if a value is not specified in `BulkWriteOptions`. This is required to
+avoid inconsistencies between server and driver behavior if the server default changes in the
+future.
+
+### Size Limits
+
+The server reports a `maxBsonObjectSize` in its `hello` response. This value defines the maximum
+size for documents that are inserted into the database. Documents that are sent to the server but
+are not intended to be inserted into the database (e.g. command documents) have a size limit of
+`maxBsonObjectSize + 16KiB`. When an acknowledged write concern is used, drivers MUST NOT perform
+any checks related to these size limits and MUST rely on the server to raise an error if a limit is
+exceeded. However, when an unacknowledged write concern is used, drivers MUST raise an error if one
+of the following limits is exceeded:
+
+- The size of a document to be inserted MUST NOT exceed `maxBsonObjectSize`. This applies to the
+  `document` field of an `InsertOneModel` and the `replacement` field of a `ReplaceOneModel`.
+- The size of an entry in the `ops` array MUST NOT exceed `maxBsonObjectSize + 16KiB`.
+- The size of the `bulkWrite` command document MUST NOT exceed `maxBsonObjectSize + 16KiB`.
+
+See [SERVER-10643](https://jira.mongodb.org/browse/SERVER-10643) for more details on these size
+limits.
+
+## Command Batching
```

Comment:
Lets specify whether `nsInfo` contains only the namespaces actually used in the current batch, or all the namespaces passed to the `bulkWrite` method (as these can be different depending on the split points).  The latter may be simpler to implement, as it makes the logic for calculating the split points more straightforward (you can pre-calculate the size of `nsInfo` before you even execute the splitting logic).

Reply:
@kevinAlbs I think C driver does the latter?

Reply:
I pushed a commit with more specific batch splitting logic plus a Q&A entry explaining the reasoning behind the design. This change should make implementation and testing of splitting up `nsInfo` simpler. @kevinAlbs, let me know if this proposed new logic makes sense for your implementation. I'm specifically curious whether differentiating between command-specific fields and command-agnostic fields is sensible. Approximating in this way will let us avoid a pretty major refactor in Rust. I will also add some tests around batch splitting on nsInfo that can rely on the 1000 bytes of overhead (i.e. the test proposed [here](https://github.com/mongodb/specifications/pull/1534#discussion_r1572908056)).